### PR TITLE
Patch to fix various issues

### PR DESCRIPTION
**Pull request by CraigLn**
* Invalid JSON will attempt to be parsed for the next Timetoken, and if successful will no interrupt the current subscription loop
* Responds with the appropriate error when a Signal payload is too large
* Added PAM support to Channels for Channel-Group endpoint
* Resolves possible memory leaks related to listeners

### CODE REVIEW

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Clones added
============
- Sources/PubNub/Networking/Response/ResponseOperator.swift  1
- Tests/PubNubTests/Networking/Routers/PublishRouterTests.swift  1
- Sources/PubNub/Networking/Routers/SubscribeRouter.swift  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=5635978&bid=18457137)

File: Sources/PubNub/Subscription/SubscriptionSession.swift [hunk 1]
```diff
@@ -316,6 +316,10 @@ public class SubscriptionSession {
               // No new request has been created so we'll reconnect here
               self?.reconnect(at: self?.previousTokenResponse?.timetoken)
             }
+          } else if let timetokenPayload = error.pubNubError?.affected.findFirst(by: PubNubError.AffectedValue.subscribe) {
```
Comment (madisonsmith) [inline Sources/PubNub/Subscription/SubscriptionSession.swift:319]: Codacy Issue found: [Line should be 120 characters or less: currently 123 characters](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=5635978)

File: Sources/PubNub/Helpers/CaseAccessible.swift [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  subscript<AssociatedValue>(case path: (AssociatedValue) -> Self) -> AssociatedValue? {
+    get { return associatedValue(matching: path) }
+  }
+
+  subscript<AssociatedValue>(
+    case path: (AssociatedValue) -> Self,
+    default value: AssociatedValue
+  ) -> AssociatedValue {
+    get { return associatedValue(matching: path) ?? value }
```
Comment (madisonsmith) [inline Sources/PubNub/Helpers/CaseAccessible.swift:66]: Codacy Issue found: [Computed read-only subscripts should avoid using the get keyword.](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=5635978)

File: Sources/PubNub/Helpers/CaseAccessible.swift [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    for case let (casePath?, value) in Mirror(reflecting: self).children {
+      if let result = (value as? AssociatedValue) ?? (Mirror(reflecting: value).children.first?.value as? AssociatedValue) {
+        return (casePath, result)
+      }
+    }
+    return nil
+  }
+
+  subscript<AssociatedValue>(case path: (AssociatedValue) -> Self) -> AssociatedValue? {
+    get { return associatedValue(matching: path) }
```
Comment (madisonsmith) [inline Sources/PubNub/Helpers/CaseAccessible.swift:59]: Codacy Issue found: [Computed read-only subscripts should avoid using the get keyword.](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=5635978)

File: Sources/PubNub/Helpers/CaseAccessible.swift [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+          let mirrorCasePath = Mirror(reflecting: casePath(decomposed.value)).children.first?.label,
+          decomposed.casePath == mirrorCasePath
+      else { return nil }
+
+    return decomposed.1
+  }
+
+  private func decompose<AssociatedValue>() -> (casePath: String, value: AssociatedValue)? {
+    for case let (casePath?, value) in Mirror(reflecting: self).children {
+      if let result = (value as? AssociatedValue) ?? (Mirror(reflecting: value).children.first?.value as? AssociatedValue) {
```
Comment (madisonsmith) [inline Sources/PubNub/Helpers/CaseAccessible.swift:51]: Codacy Issue found: [Line should be 120 characters or less: currently 124 characters](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=5635978)

File: Tests/PubNubTests/Networking/Routers/SubscribeRouterTests.swift [hunk 5]
```diff
@@ -774,6 +774,57 @@ extension SubscribeRouterTests {
   }
 }
 
+// MARK: - Error Handling
+extension SubscribeRouterTests {
+  func testInvalidJSONResponse() {
+    let corruptBase64Response = "eyJ0Ijp7InQiOiIxNTkxMjE4MzQ0MTUyNjM1MCIsInIiOjF9LCJtIjpbeyJhIjoiMyIsImYiOjUxMiwicCI6eyJ0IjoiMTU5MTIxODM0NDE1NTQyMDAiLCJyIjoxfSwiayI6ImRlbW8tMzYiLCJjIjoic3dpZnRJbnZhbGlkSlNPTi7/IiwiZCI6ImhlbGxvIiwiYiI6InN3aWZ0SW52YWxpZEpTT04uKiJ9XX0="
```
Comment (madisonsmith) [inline Tests/PubNubTests/Networking/Routers/SubscribeRouterTests.swift:780]: Codacy Issue found: [Line should be 120 characters or less: currently 266 characters](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=5635978)

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Sources/PubNub/Networking/Response/ResponseOperator.swift  1
- Tests/PubNubTests/Networking/Routers/PublishRouterTests.swift  1
- Sources/PubNub/Networking/Routers/SubscribeRouter.swift  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=5635978&bid=18457137)

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Sources/PubNub/Networking/Response/ResponseOperator.swift  1
- Tests/PubNubTests/Networking/Routers/PublishRouterTests.swift  1
- Sources/PubNub/Networking/Routers/SubscribeRouter.swift  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=5635978&bid=18457137)

File: Sources/PubNub/Networking/Routers/SubscribeRouter.swift [hunk 6]
```diff
@@ -106,6 +106,37 @@ struct SubscribeRouter: HTTPRouter {
...
+      let decodedResponse = EndpointResponse<Payload>(router: response.router,
+                                                      request: response.request,
+                                                      response: response.response,
+                                                      data: response.data,
+                                                      payload: decodedPayload)
+
+      return .success(decodedResponse)
+    } catch {
+      // Atempt to parse out the timetoken protion of the payload to push ahead the subscribe loop
+      if response.payload.count >= 37 {
```
Comment (davidnub) [inline Sources/PubNub/Networking/Routers/SubscribeRouter.swift:122]: Are we guaranteed this length to be fixed?

Comment (CraigLn) [reply Sources/PubNub/Networking/Routers/SubscribeRouter.swift:122]: From what I've read JSON data is guaranteed to be ordered, and the goal of this code is to pull out the time payload from the response.  

The only variance would be whether the Timetoken or Region Int byte size were to change, or the system starts sending the time payload after the message portion, and neither of those are really controlled by the client so this is really just a better-effort attempt to keep the loop going. Worst case this fails, and then they're no worse off than they were previously (i.e. They get an error that they couldn't continue, and their loop stops)

Comment (davidnub) [reply Sources/PubNub/Networking/Routers/SubscribeRouter.swift:122]: Ok thanks for the explanation. Sounds logical enough to use this method until we can't.

**Comment by madisonsmith** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- Sources/PubNub/Networking/Routers/SubscribeRouter.swift  1
- Tests/PubNubTests/Networking/Routers/PublishRouterTests.swift  1
- Sources/PubNub/Networking/Response/ResponseOperator.swift  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/PubNub/swift/pullRequest?prid=5635978&bid=18457137)

**Pull request merged**